test/common/config/utility_test.cc

```diff
@@ -295,6 +296,65 @@ TEST(UtilityTest, PrepareDnsRefreshStrategy) {
   }
 }
 
+void packTypedStructIntoAny(ProtobufWkt::Any& typed_config, const ProtobufWkt::Message& inner) {
+  udpa::type::v1::TypedStruct typed_struct;
+  (*typed_struct.mutable_type_url()) =
+      std::string("type.googleapis.com/") + inner.GetDescriptor()->full_name();
+  MessageUtil::jsonConvert(inner, *typed_struct.mutable_value());
+  typed_config.PackFrom(typed_struct);
+}
+
+TEST(UtilityTest, TypedStructToStruct) {
+  ProtobufWkt::Any typed_config;
+  ProtobufWkt::Struct untyped_struct;
+  (*untyped_struct.mutable_fields())["foo"].set_string_value("bar");
+  packTypedStructIntoAny(typed_config, untyped_struct);
+
+  ProtobufWkt::Struct out;
+  Utility::translateOpaqueConfig(typed_config, ProtobufWkt::Struct(),
+                                 ProtobufMessage::getStrictValidationVisitor(), out);
+
+  EXPECT_TRUE(Protobuf::util::MessageDifferencer::Equals(out, untyped_struct));
+}
+
+TEST(UtilityTest, TypedStructToBootstrap) {
+  ProtobufWkt::Any typed_config;
+  envoy::config::bootstrap::v2::Bootstrap bootstrap;
+  const std::string bootstrap_config_yaml = R"EOF(
+    admin:
+      access_log_path: /dev/null
+      address:
+        pipe:
+          path: "/"
+  )EOF";
+  TestUtility::loadFromYaml(bootstrap_config_yaml, bootstrap);
+  packTypedStructIntoAny(typed_config, bootstrap);
+
+  envoy::config::bootstrap::v2::Bootstrap out;
+  Utility::translateOpaqueConfig(typed_config, ProtobufWkt::Struct(),
+                                 ProtobufMessage::getStrictValidationVisitor(), out);
+  EXPECT_TRUE(Protobuf::util::MessageDifferencer::Equals(out, bootstrap));
+}
+
+TEST(UtilityTest, TypedStructToInvalidType) {
+  ProtobufWkt::Any typed_config;
+  envoy::config::bootstrap::v2::Bootstrap bootstrap;
+  const std::string bootstrap_config_yaml = R"EOF(
+    admin:
+      access_log_path: /dev/null
+      address:
+        pipe:
+          path: "/"
+  )EOF";
+  TestUtility::loadFromYaml(bootstrap_config_yaml, bootstrap);
+  packTypedStructIntoAny(typed_config, bootstrap);
+
+  ProtobufWkt::Any out;
+  EXPECT_THROW(Utility::translateOpaqueConfig(typed_config, ProtobufWkt::Struct(),
```

Comment:
Can you expect on the more specific message that this throws?

Reply:
done